compiler/src/Language/Bond/Codegen/Cpp/Grpc_h.hs

```diff
@@ -141,10 +144,68 @@ inline #{className}::#{proxyName}<TThreadPool>::#{proxyName}(
     { }
 
 #{doubleLineSep 0 methodDecl serviceMethods}
+
+#{template}struct #{className}::Schema
+{
+    #{export_attr}static const ::bond::Metadata metadata;
+
+    #{newlineSep 1 methodMetadata serviceMethods}
+
+    public: struct service
+    {
+        #{doubleLineSep 2 methodTemplate serviceMethods}
+    };
+
+    private: typedef boost::mpl::list<> methods0;
+    #{newlineSep 1 pushMethod indexedMethods}
+
+    public: typedef #{typename}methods#{length serviceMethods}::type methods;
+
+    #{constructor}
+};
+#{onlyTemplate $ CPP.schemaMetadata cpp s}
 |]
       where
         className = CPP.className s
         template = CPP.template s
+        onlyTemplate x = if null declParams then mempty else x
+        typename = onlyTemplate [lt|typename |]
+
+        export_attr = optional (\a -> [lt|#{a}
+        |]) export_attribute
+
+        methodMetadataVar m = [lt|s_#{methodName m}_metadata|]
+
+        methodMetadata m =
+            [lt|private: #{export_attr}static const ::bond::Metadata #{methodMetadataVar m};|]
+
+        -- reversed list of method names zipped with indexes
+        indexedMethods :: [(String, Int)]
+        indexedMethods = zipWith ((,) . methodName) (reverse serviceMethods) [0..]
+
+        pushMethod (method, i) =
+            [lt|private: typedef #{typename}boost::mpl::push_front<methods#{i}, #{typename}service::#{method}>::type methods#{i + 1};|]
+
+        -- constructor, generated only for service templates
+        constructor = onlyTemplate [lt|Schema()
+        {
+            // Force instantiation of template statics
+            (void)metadata;
+            #{newlineSep 4 static serviceMethods}
```

Comment:
Minor: The indentation is off on these. I think we want an indentation level of 3, not 4. (See the golden files.)